File: solutions/elixir/027-search-in-rotated-sorted-array.ex
# Problem 27: Search in Rotated Sorted Array (LeetCode 33)
# Difficulty: Med
# Language: Elixir
# 
defmodule SearchRotated do
  def search(nums, target) do
    arr = :array.from_list(nums)
    go(arr, target, 0, :array.size(arr) - 1)
  end

  defp go(_arr, _target, lo, hi) when lo > hi, do: -1
  defp go(arr, target, lo, hi) do
    mid = div(lo + hi, 2)
    mid_val = :array.get(mid, arr)
    lo_val = :array.get(lo, arr)
    hi_val = :array.get(hi, arr)
    cond do
      mid_val == target -> mid
      lo_val <= mid_val ->
        if target >= lo_val and target < mid_val,
          do: go(arr, target, lo, mid - 1),
          else: go(arr, target, mid + 1, hi)
      true ->
        if target > mid_val and target <= hi_val,
          do: go(arr, target, mid + 1, hi),
          else: go(arr, target, lo, mid - 1)
    end
  end
end

File: solutions/ocaml/027-search-in-rotated-sorted-array.ml
(*
  Problem 27: Search in Rotated Sorted Array (LeetCode 33)
  Difficulty: Med
  Language: OCaml
*)
let search nums target =
  let a = Array.of_list nums in
  let rec go lo hi =
    if lo > hi then -1
    else
      let mid = (lo + hi) / 2 in
      if a.(mid) = target then mid
      else if a.(lo) <= a.(mid) then
        if target >= a.(lo) && target < a.(mid) then go lo (mid - 1)
        else go (mid + 1) hi
      else
        if target > a.(mid) && target <= a.(hi) then go (mid + 1) hi
        else go lo (mid - 1)
  in
  go 0 (Array.length a - 1)
